### PR TITLE
allow generating during build by building from an output dir

### PR DESCRIPTION
**Pull request by Alan-Chen99**
solves https://github.com/danielbarter/mini_compile_commands/issues/5

### CODE REVIEW

File: .nixd.json
Comment (danielbarter) [inline .nixd.json:1]: This shouldn't be version controlled

File: README.md [hunk 1]
```diff
@@ -51,14 +51,25 @@ There are tests for `gcc` and `clang` in `tests/gcc` and `tests/clang` respectiv
 
 ## Build input hook
 
-It is possible to use mini compile commands to generate a `compile_commands.json` while building a derivation. This involves using a wrapped standard standard environment and adding a hook to the `buildInputs`:
-```
-with import <nixpkgs> {};
-let mcc-env = (callPackage <this_repo> {}).wrap stdenv;
-    mcc-hook = (callPackage <this_repo> {}).hook;
-in (hello.override { stdenv = mcc-env; }).overrideAttrs
+It is possible to use mini compile commands to generate a `compile_commands.json` while building a derivation. This works by building directly in a subdirectory of `$out` or some other `$custom-output` (so that `compile_commands.json` contains valid paths).
```
Comment (danielbarter) [inline README.md:54]: This seems incredibly suspicious to me. We don't want to interfere with the natural build process in any way, other than to pull information out of the compiler wrapper. For example, what if the standard build directory is on a fast, local file system and the nix store is on some remote, slow file system? This will almost surely unearth a lot of build time concurrency bugs

File: README.md [hunk 2]
```diff
@@ -51,14 +51,25 @@ There are tests for `gcc` and `clang` in `tests/gcc` and `tests/clang` respectiv
...
+
+    # can also put in another output
+    # MCC_BUILD_DIR = "$dev";
+    # outputs = ["out" "dev"];
   })
 ```
-Running `nix-build` on this derivation will generate a `compile_commands.json` and move it into `$out`. As explained in [the compile commands specification](https://clang.llvm.org/docs/JSONCompilationDatabase.html), each entry contains a `directory` attribute which is the absolute path of the directory from which the compiler invocation occurred. To use this generated json with `clangd`, the source code needs to be located in the same place as during the nix build (typically `/build`). Since this is inconvenient, we recommend using mini compile commands interactively.
+
+Running `nix-build` on this derivation will generate a `compile_commands.json` at `$MCC_BUILD_DIR/compile_commands.json`.
+by default all files other than c and header files are removed in `$MCC_BUILD_DIR`. This can be customized by setting `MCC_KEEP_REGEXP`, which is any python regex.
```
Comment (danielbarter) [inline README.md:75]: This whole approach is making no sense to me. If I am understanding correctly, you are copying the source into the nix store, and then building in the store, but the store is read only. How does this even work for builds that modify their source files, like the C preprocessor? 

Also, at the end, you are going to be left with a potentially incomplete copy of the source tree, with a compile commands json that is on a read only file system, so you won't be able to change any of the code or actually run a build?

File: default.nix [hunk 3]
```diff
@@ -15,34 +15,48 @@ rec {
       # be split out: https://github.com/NixOS/nixpkgs/issues/132340. Once that happens,
       # this hack will no longer be needed (and will no longer work)
       gcc-hack = lib.optionalString env.cc.isGNU ''
-           libcxxflags=$out/nix-support/libcxx-cxxflags
-           libcflags=$out/nix-support/libc-cflags
-           echo -isystem ${env.cc.cc}/include/c++/${env.cc.cc.version} >> $libcxxflags
-           echo -isystem ${env.cc.cc}/include/c++/${env.cc.cc.version}/${env.hostPlatform.config} >> $libcxxflags
-           echo -isystem ${env.cc.cc}/lib/gcc/${env.hostPlatform.config}/${env.cc.cc.version}/include >> $libcflags
+        libcxxflags=$out/nix-support/libcxx-cxxflags
```
Comment (danielbarter) [inline default.nix:18]: You need to split linting out into a separate PR, otherwise I need to spend a lot of effort deciding what is a functional change and what is a cosmetic change

File: flake.nix
Comment (danielbarter) [inline flake.nix:1]: What is the purpose of adding a flake with all these dependencies?

File: mini_compile_commands_clean.py
Comment (danielbarter) [inline mini_compile_commands_clean.py:1]: This seems weird to me. Why would you ever want to have editor integration in a pruned source tree?

File: nixd-entry.nix
Comment (danielbarter) [inline nixd-entry.nix:1]: Why?

**Comment by danielbarter** [comment]
@Alan-Chen99: I very much appreciate you taking a shot at this, but I am pretty skeptical.

Changing where a build happens is introducing timing impurities into the build process (the standard build directory might be on a fast file system, and the store might be on a slow file system), and unfortunately, building software is riddled with concurrency bugs that will be inevitably triggered by such a change.

Moreover, from a developer perspective, it is very hard to see how these hook based approaches are useful. Why would I ever want to have a `compile_commands.json` for a source tree, stored on a read only file system? The whole point is that I want lsp to work for a source tree that I am actively working on. 

Also, you have mixed in a bunch of linting and non essential files (nixd stuff, the flake) that make the PR harder to review. This is really a mix of 3 PRs, which you need to split apart.


I think the correct solution here is is that `mini_compile_commands_server.py` should generate a `compile_commands.json` parameterized over the build directory (for example, this could be a nix function). This would make it relocatable, and you could evaluate it for wherever you end up storing your source tree.

**Comment by Alan-Chen99** [comment]
sorry for not being clear.

I made this to run on library dependencies, and interpreters (python, emacs, etc), since I find that I often need to browser their source code, but I dont intend to actually modify them.

I actually tried to make a relocatable solution, but this basically never work (since you often have generated files like config.h, etc)

File: README.md (hunk 1 above)
Comment (Alan-Chen99) [reply README.md:54]: when building nix uses a mount on /tmp, so this is still running on the local file system. after the build, if successful, the result is copied to the actual store

File: flake.nix
Comment (Alan-Chen99) [reply flake.nix:1]: I find that [nixd](https://github.com/nix-community/nixd) is very useful, and it would be nice if more people used it, but it takes some work to setup and one have to modify some files

File: default.nix (hunk 3 above)
Comment (Alan-Chen99) [reply default.nix:18]: will fix

File: README.md (hunk 1 above)
Comment (danielbarter) [reply README.md:54]: Ah OK, this makes sense.

**Comment by danielbarter** [comment]
Hi @Alan-Chen99: I am gonna close because I think this adds too much complexity for a feature which I don't really understand, and personally don't find useful. 

You are more than welcome to maintain a fork, or drastically simplify the MR.
